### ISSUE TITLE
patch modules doesn't source from files path when using roles

### ISSUE THREAD

**Issue by ansibot** [ansible/ansible#29203]
From @kputland on 2015-04-29T22:14:28Z
##### ISSUE TYPE
Bug Report

##### COMPONENT NAME
patch module

##### ANSIBLE VERSION
2.0

##### SUMMARY
patch modules doesn't source from files path when using roles

TASK: [elasticsearch | tweak service] ****************************************\* 
failed: [172.16.16.6] => {"failed": true}
msg: src elasticsearch.service.patch doesn't exist or not readable
failed: [172.16.16.5] => {"failed": true}
msg: src elasticsearch.service.patch doesn't exist or not readable
failed: [172.16.16.3] => {"failed": true}
msg: src elasticsearch.service.patch doesn't exist or not readable
failed: [172.16.16.4] => {"failed": true}
msg: src elasticsearch.service.patch doesn't exist or not readable
failed: [172.16.16.2] => {"failed": true}
msg: src elasticsearch.service.patch doesn't exist or not readable

FATAL: all hosts have already failed -- aborting

PLAY RECAP *******************************************************************\* 
           to retry, use: --limit @/home/karl/elasticsearch.retry

172.16.16.2                : ok=15   changed=8    unreachable=0    failed=1  
172.16.16.3                : ok=15   changed=8    unreachable=0    failed=1  
172.16.16.4                : ok=15   changed=8    unreachable=0    failed=1  
172.16.16.5                : ok=15   changed=8    unreachable=0    failed=1  
172.16.16.6                : ok=15   changed=8    unreachable=0    failed=1   

[karl@node1 ansible]$ ls roles/elasticsearch/files/elasticsearch.service.patch 
roles/elasticsearch/files/elasticsearch.service.patch

Copied from original issue: ansible/ansible-modules-extras#434

**Comment by ansibot** [comment]
From @kputland on 2015-04-29T22:14:28Z

I can confirm this bug.

Error:

```
TASK: [ambari-server | patch ambari-server to work arround systemd incompatibilities] ***
failed: [ambari] => {"failed": true}
msg: src systemd.patch doesn't exist or not readable
```

Role: 

```
$ cat roles/ambari-server/tasks/main.yml
---
- name: install ambari-server
  yum: name=ambari-server state=latest

- name: ensure PostgreSQL is enabled
  service: name=postgresql enabled=yes

- name: patch ambari-server to work arround systemd incompatibilities
  patch: src=systemd.patch basedir=/usr/lib/python2.6/site-packages/ambari_server/
```

... but file exists:

```
$ ls roles/ambari-server/files/
systemd.patch
```

**Comment by ansibot** [comment]
From @kputland on 2015-04-29T22:14:28Z

Yes, what is needed is an action plugin to accompany this module.  [Nicolas Grisey Demengel has a good write-up of what is required](https://ndemengel.github.io/2015/01/20/ansible-modules-and-action-plugins/#action-plugins).  He says:

> As explained on <a href="https://groups.google.com/forum/#!searchin/ansible-project/action$20plugin$20vs$20module/ansible-project/wNSJ4g-f4F8/xlpijLcRNSUJ">Ansible’s Google Group</a>: “action_plugins are a special type of module, or a compliment to existing modules. action_plugins get run on the ‘master’ instead of on the target, for modules like file/copy/template, some of the work needs to be done on the master before it executes things on the target. The action plugin executes first and can then execute (or not) the normal module”.<br>
> For instance, the <a href="https://github.com/ansible/ansible/blob/devel/lib/ansible/runner/action_plugins/copy.py">copy action plugin</a> calls (<code>_execute_module</code>) the <a href="https://github.com/ansible/ansible-modules-core/blob/devel/files/copy.py">copy module</a>.

Where I am stuck is where to put the patch action plugin in the [ansible/ansible-modules-extras](https://github.com/ansible/ansible-modules-extras) file heirarchy.  There doesn't seem to be another module inside ansible-modules-extras to use as guidance.

**Comment by ansibot** [comment]
From @kputland on 2015-04-29T22:14:28Z

So the issue here is that action plugins currently live in ansible/ansible. Which means that you'd need to submit a PR for the action plugin to that repository currently.

**Comment by ansibot** [comment]
From @kputland on 2015-04-29T22:14:28Z

[module: files/patch.py]

ping module maintainer @jirutka for additional guidance here.

**Comment by ansibot** [comment]
From @kputland on 2015-04-29T22:14:28Z

confirm. is not working with files in roles.

**Comment by ansibot** [comment]
From @kputland on 2015-04-29T22:14:28Z

Can we fix the documentation to indicate this? The src has to be copied to the target before using the patch module.

**Comment by ansibot** [comment]
From @kputland on 2015-04-29T22:14:28Z

there IS a patch action plugin, this should 'just work'TM. By default it should look in the role's files/ dir and  follow the normal search path after that (role_dir/tasks/, role_dir/, playdir/files, playdir/)

**Comment by ansibot** [comment]
From @kputland on 2015-04-29T22:14:28Z

@jirutka, @luisperlaz, ping. This issue is still waiting on your response.
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by ansibot** [comment]
From @kputland on 2015-04-29T22:14:28Z

@jirutka, @luisperlaz, ping. This issue is still waiting on your response.
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by ansibot** [comment]
From @kputland on 2015-04-29T22:14:28Z

@jirutka, @luisperlaz, ping. This issue is still waiting on your response.
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by ansibot** [comment]
From @kputland on 2015-04-29T22:14:28Z

@jirutka, @luisperlaz, ping. This issue is still waiting on your response.
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by ansibot** [comment]
From @kputland on 2015-04-29T22:14:28Z

@jirutka, @luisperlaz, ping. This issue is still waiting on your response.
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by ansibot** [comment]
From @kputland on 2015-04-29T22:14:28Z

@jirutka, @luisperlaz, ping. This issue is still waiting on your response.
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by ansibot** [comment]
@ansibot Greetings! Thanks for taking the time to open this issue. In order for the community to handle your issue effectively, we need a bit more information. 

Here are the items we could not find in your description:
- issue type
- ansible version
- component name

Please set the description of this issue with this template:
https://raw.githubusercontent.com/ansible/ansible/devel/.github/ISSUE_TEMPLATE.md

[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by traumschule** [comment]
> Here are the items we could not find in your description:
> 
>     issue type
>     ansible version
>     component name

To me it looks like all three are specified above.

**Comment by sivel** [comment]
This issue is no longer present in recent versions of Ansible. If you are still having problems with a recent version of Ansible, please feel free to open a new issue.

If you have further questions please stop by IRC or the mailing list:

   * IRC: #ansible on irc.freenode.net
   * mailing list: https://groups.google.com/forum/#!forum/ansible-project
